nft_ingester/src/program_transformers/bubblegum/transfer.rs

```diff
@@ -21,29 +22,43 @@ where
 {
     if let (Some(le), Some(cl)) = (&parsing_result.leaf_update, &parsing_result.tree_update) {
         let seq = save_changelog_event(cl, bundle.slot, txn).await?;
+        #[allow(unreachable_patterns)]
         return match le.schema {
             LeafSchema::V1 {
                 id,
-                delegate,
                 owner,
+                delegate,
                 ..
             } => {
-                let id_bytes = id.to_bytes().to_vec();
+                let id_bytes = id.to_bytes();
+                let owner_bytes = owner.to_bytes().to_vec();
                 let delegate = if owner == delegate {
                     None
                 } else {
                     Some(delegate.to_bytes().to_vec())
                 };
-                let owner_bytes = owner.to_bytes().to_vec();
-                let asset_to_update = asset::ActiveModel {
-                    id: Unchanged(id_bytes.clone()),
-                    leaf: Set(Some(le.leaf_hash.to_vec())),
-                    delegate: Set(delegate),
-                    owner: Set(Some(owner_bytes)),
-                    seq: Set(seq as i64), // gummyroll seq
-                    ..Default::default()
-                };
-                update_asset(txn, id_bytes, Some(seq), asset_to_update).await
+
+                // Partial update of asset table with just leaf.
+                upsert_asset_with_leaf_info(
+                    txn,
+                    id_bytes.to_vec(),
+                    Some(le.leaf_hash.to_vec()),
+                    Some(seq as i64),
+                    false,
+                )
+                .await?;
+
+                // Partial update of asset table with just leaf owner and delegate.
+                upsert_asset_with_owner_and_delegate_info(
+                    txn,
+                    id_bytes.to_vec(),
+                    owner_bytes,
+                    delegate,
+                    seq as i64,
+                )
+                .await?;
+
+                upsert_asset_with_seq(txn, id_bytes.to_vec(), seq as i64).await
```

Comment:
Is this necessary? I think we can just update this within one of the earlier methods. We try to reduce to number of DB calls we need to make in favour of performance. (applies for all cNFT indexing methods).

Reply:
In a previous iteration of this PR I used the existing `seq` to protect `leaf` updates, and the new `owner_delegate_seq` to protect `owner` and `delegate` fields. Because of this, an effect of my PR changes is that when the leaf was set to `None` during decompression, so was the `seq`, and then for decompressed assets, a Read API call to `getAsset` would return `null` for the `seq`. 

Maybe that was OK because people don't need the `seq` when the asset has been removed from the merkle tree. But the merkle tree is still listed and wanted it to match previous behavior.

Maybe I can change it back if we want to remove an extra update that is only for the `seq`.

Reply:
> a Read API call to getAsset would return null for the seq.
This would be acceptable if the asset can never be re-compresssed. But I think you plan on adding that, right?

If that's the case, then it makes sense to keep the seq value available after a decompression.

Reply:
Yeah I think long term we plan to add re-compression.  We have the placeholder for re-compression but not fully designed yet.  Maybe it would insert back into the same tree like `cancel_redeem` does but instead of using a voucher it would burn the token-metadata based NFT.  But I could also see it just doing a new append to the tree via `mint_v1`.

Either way when it got added back into the tree, I think the asset would get a new `seq` that is probably a much higher number that what was previously stored there when it was decompressed, as many transfers and other operations probably happened on the tree while this NFT was decompressed.

I guess what I'm thinking at the moment is that I'm not sure of the use for the `seq` after decompression since you cannot use it for re-compression, right?  Or am I misunderstanding something?

Reply:
For reference the previous version of my PR prior to [this commit](https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/pull/77/commits/f7a51430cef898f76cebd3f3aaaa74a9eff3c903) used `seq` to protect leaf and `owner_delegate_seq` to protect `owner`+`delegate` and had one less upsert overall.  May be worth revisiting that approach to save on an update/upsert.

Reply:
> I guess what I'm thinking at the moment is that I'm not sure of the use for the seq after decompression since you cannot use it for re-compression, right?

If recompressing will add that NFT back into the tree via the same leaf, then we need to keep the seq number around. Otherwise you can have out of order bugs with out of order decompress + recompress transactions. And we need to be able to always handle any order to be safe, even if its unlikely to happen.

Reply:
I think you uncovered something that will be an issue for `decompress` > `recompress` with the current design both of this PR and with the system in general.  The fact that decompress doesn't have any association with the tree, means there is no change log event and thus no `seq` associated with the `decompress`.

For this PR, the way I found to deal with `decompress` being out of order with others was to add a boolean `was_decompressed` flag, which essentially acts like a 1-bit sequence number, which is a one way door assuming no events after the decompression.

Reply:
I still don't think recompression can work out of order, because decompression is not sequenced by the tree.  Am I missing something?